Add pre-commit check for cli.markdown updates and coffeelint execution

pdcastro opened the pull request:
This PR introduces a pre-commit file timestamp comparison between `doc/cli.markdown` (CLI markdown documentation) and any staged files under the `lib` and `typings` folders. If `cli.markdown` is older, an error is printed and the commit aborted:
```
$ git commit -a
husky > pre-commit (node v10.18.1)
--------------------------------------------------------------------------------
ERROR: at least one staged file: "lib/utils/promote.ts"
has a more recent modification timestamp than the documentation file:
"doc/cli.markdown"

This probably means that `npm run build` or `npm test` have not been executed,
and this error can be fixed by doing so. Running `npm run build` or `npm test`
before commiting is currently a requirement (documented in the CONTRIBUTING.md
file) for three reasons:
1. Because it updates the CLI markdown documentation (in case any command-line
   options were updated, added or removed).
2. Because it reveals coffeelint warnings that the CI would currently ignore
   (in case any Coffeescript files were modified).
3. Given that balena-cli CI builds/tests are currently fairly slow, running
   tsc or resin-lint on the laptop catches Typescript type check errors much
   sooner and typically results in reduced overall waiting time.

If you need/wish to bypass this check without running `npm run build`, run
`npx touch -am "doc/cli.markdown"` and try again.
--------------------------------------------------------------------------------
husky > pre-commit hook failed (add --no-verify to bypass)
```

Change-type: patch

pdcastro commented on automation/check-doc.js:
```diff
@@ -0,0 +1,90 @@
+
+				If you need/wish to bypass this check without running \`npm run build\`, run
+				\`npx touch -am "${docFile}"\` and try again.
+				--------------------------------------------------------------------------------
```
The reason for advising `npx touch` instead of simply `touch` is Windows compatibility (`touch` is a Unix utility).

pdcastro commented on package.json:
```diff
@@ -148,6 +148,7 @@
     "publish-release": "^1.6.1",
     "resin-lint": "^3.2.4",
     "rewire": "^3.0.2",
+    "simple-git": "^1.131.0",
```
[simple-git](https://www.npmjs.com/package/simple-git) is a lightweight wrapper around executing `git`, conveniently parsing the output as objects and arrays. The `git` executable must still be available.

pdcastro commented on automation/check-doc.js:
```diff
@@ -0,0 +1,90 @@
+	const docFile = path.join('doc', 'cli.markdown');
+	const [docStat, gitStatus] = await Promise.all([
+		fs.stat(docFile),
+		git.status(),
```
`git.status()` provides similar information as running `git status` on a command line, with the result returned in the form of objects and arrays. Here it is used to identify staged files (including newly created or renamed files).

I had originally intended to use `lint-staged` (as I had suggested in some comments in PR #1589) instead of anything like `git status`, because `lint-staged`'s purpose is exactly to run scripts with staged files as arguments. Alas, I found that `lint-staged` modifies the timestamps of staged files, which completely foiled my timestamp comparison plan! No matter: `git.status()` to the rescue. It should be OK for as long as this script is executed **before** `lint-staged`.

pdcastro commented on package.json:
```diff
@@ -52,7 +52,7 @@
     "pretest": "npm run build",
     "test": "mocha --timeout 6000 -r ts-node/register/transpile-only \"tests/**/*.spec.ts\"",
-    "test:fast": "npm run build:fast && npm run test",
+    "test:fast": "npm run build:fast && mocha --timeout 6000 -r ts-node/register/transpile-only \"tests/**/*.spec.ts\"",
```
This modification is in its own commit because it is an independent fix. What happens is that `npm test` triggers the `pretest` script that executes `npm run build`. Before this change, `npm run test:fast` would run `npm run build:fast` but then also run `npm run build` (because of `prestest`), therefore building twice and being slower than `npm test`.

pdcastro commented:
@balena-ci retest

srlowe commented:
@balena-ci retest

hedss commented on automation/check-doc.js:
```diff
@@ -0,0 +1,90 @@
+		await checkBuildTimestamps();
+	} catch (err) {
+		console.error(err.message);
+		process.exit(1);
```
```suggestion
		process.exitCode = 1;
```
Whilst it doesn't usually happen, `libuv` can still be holding onto file handles after it's accessed files, and ensuring that it cleans itself up before exiting at the next event loop just makes things a bit better. `process.exitCode` ensures that the process still exists non-zero, but allows it to run the rest of the codepath in the current event loop and clean up first (and looking at the calling convention here, this is fine).

pdcastro commented:
Force push:
* Replaced `process.exit(1)` with `process.exitCode = 1` ([Hedley's comment](https://github.com/balena-io/balena-cli/pull/1623#discussion_r380852211)).
* Added the project root folder as an argument to simple-git (now `simplegit(ROOT)`, previously `simplegit()`) so that the script can be executed from any current working dir, not just the project root.

pdcastro commented:
Force pushed to update the PR branch (rebase on master). I had not merged the PR before because I wanted to run some additional tests on Windows. Tested now, looks good, will merge, thanks for the reviews! 👍
